src/python/pants/jvm/resolve/coursier_fetch.py

```diff
@@ -0,0 +1,526 @@
+            dependencies=[coord.coord for coord in self.dependencies],
+            file_name=self.file_name,
+            file_digest=dict(
+                fingerprint=self.file_digest.fingerprint,
+                serialized_bytes_length=self.file_digest.serialized_bytes_length,
+            ),
+        )
+
+
+@dataclass(frozen=True)
+class CoursierResolvedLockfile:
+    """An in-memory representation of Pants' Coursier lockfile format."""
+
+    entries: Tuple[CoursierLockfileEntry, ...]
+
+    @classmethod
+    def from_json_dict(cls, lockfile) -> CoursierResolvedLockfile:
+        """Construct a CoursierResolvedLockfile from its JSON dictionary representation."""
+
+        return CoursierResolvedLockfile(
+            entries=tuple(CoursierLockfileEntry.from_json_dict(dep) for dep in lockfile)
+        )
+
+    def to_json(self) -> bytes:
+        """Export this CoursierResolvedLockfile to human-readable JSON.
+
+        This JSON is intended to be checked in to the user's repo as a hermetic snapshot of a
+        Coursier resolved JVM classpath.
+        """
+        return json.dumps([entry.to_json_dict() for entry in self.entries], indent=4).encode(
+            "utf-8"
+        )
+
+
+@rule(level=LogLevel.DEBUG)
+async def coursier_resolve_lockfile(
+    coursier: Coursier,
+    maven_requirements: MavenRequirements,
+) -> CoursierResolvedLockfile:
+    """Run `coursier fetch ...` against a list of Maven coordinates and capture the result.
+
+    This rule does two things in a single Process invocation:
+
+        * Runs `coursier fetch` to let Coursier do the heavy lifting of resolving
+          dependencies and downloading resolved artifacts (jars, etc).
+        * Copies the resolved artifacts into the Process output directory, capturing
+          the artifacts as content-addressed `Digest`s.
+
+    It's important that this happens in the same process, since the process isn't
+    guaranteed to run on the same machine as the rule, nor is a subsequent process
+    invocation.  This guarantees that whatever Coursier resolved, it was fully
+    captured into Pants' content addressed artifact storage.
+
+    Note however that we still get the benefit of Coursier's "global" cache if it
+    had already been run on the machine where the `coursier fetch` runs, so rerunning
+    `coursier fetch` tends to be fast in practice.
+
+    Finally, this rule bundles up the result into a `CoursierResolvedLockfile`.  This
+    data structure encapsulates everything necessary to either materialize the
+    resolved dependencies to a classpath for Java invocations, or to write the
+    lockfile out to the workspace to hermetically freeze the result of the resolve.
+    """
+
+    if len(maven_requirements) == 0:
+        return CoursierResolvedLockfile(entries=())
+
+    coursier_report_file_name = "coursier_report.json"
+    process_result = await Get(
+        ProcessResult,
+        Process(
+            argv=[
+                "/bin/sh",
```

Comment:
Instead, ask for the `BashBinary` type from `pants.engine.process` in your rule signature. It's not as safe to assume `/bin/sh` exists.